File: 2021/02/answer.rb
def move_1((x, y), (direction, distance)) # dimensions, command)
  case direction
  when "forward" then [x + distance, y]
  when "down" then [x, y + distance]
  when "up" then [x, y - distance]
  end
end

def move_2((x, y, aim), (direction, distance)) # dimensions, command)
  case direction
  when "forward" then [x + distance, y + (distance * aim), aim]
  when "down" then [x, y, aim + distance]
  when "up" then [x, y, aim - distance]
  end
end

inputs = File.read("input.txt") # 2086357770
  .split
  .each_slice(2)
  .map { |direction, distance| [direction, distance.to_i] }

# part 1
x1, y1 = inputs
  .reduce([0, 0]) { |acc, input| move_1(acc, input) }

puts x1, y1
puts x1 * y1

# part 2
x2, y2 = inputs
  .reduce([0, 0, 0]) { |acc, input| move_2(acc, input) }

puts x2, y2
puts x2 * y2

File: 2021/02/answer.exs
defmodule App do
  def move_1([direction, distance], [x, y]) do
    case direction do
      "forward" -> [x + distance, y]
      "down" -> [x, y + distance]
      "up" -> [x, y - distance]
    end
  end

  def move_2([direction, distance], [x, y, aim]) do
    case direction do
      "forward" -> [x + distance, y + (distance * aim), aim]
      "down" -> [x, y, aim + distance]
      "up" -> [x, y, aim - distance]
    end
  end
end

inputs = File.read!("input.txt")
# inputs = """
# forward 5
# down 5
# forward 8
# up 3
# down 8
# forward 2
# """ # 150 & 900
  |> String.split
  |> Enum.chunk_every(2)
  |> Enum.map(fn [direction, distance] -> [direction, String.to_integer(distance)] end)

# part 1
[x_1, y_1] = inputs
  |> Enum.reduce([0, 0], &App.move_1/2)

IO.puts("#{x_1} #{y_1}")
IO.puts(x_1 * y_1)

# part 2
[x_2, y_2, _] = inputs
  |> Enum.reduce([0, 0, 0], &App.move_2/2)

IO.puts("#{x_2} #{y_2}")
IO.puts(x_2 * y_2)
